### ISSUE TITLE
CF template timeout on AWS::AutoScaling::AutoScalingGroup step for existing VPC

### ISSUE THREAD

**Issue by strofimo** [aws-quickstart/quickstart-linux-bastion#16]
Hi,

I tried to use this CF template to setup linux bastion in existing VPC. 
Unfortunately  template times out on "AWS::AutoScaling::AutoScalingGroup" step each time I tried. 
When it failed I see that EC2 and EIP resources created successfully but EIP is not associated with EC2 instance. If I associate it manually I'm able to SSH to EC2 instance without any issues. I also checked that instance have access to internet. But I don't see the "bastion_bootstrap.sh" in temp folder. Any idea whats wrong? Is something wrong with "AWS::CloudFormation::Init" step so "bastion_bootstrap.sh" failed to download on instance?

thanks

**Comment by strofimo** [comment]
I also checked /var/log and don't see any cfn-init logs in it:(

**Comment by yarrowfish** [comment]
Hi!  

Do you have  "Enable auto-assign public IPv4 address" turned on for your Public Subnets were the Bastions are?  

/Riad

**Comment by strofimo** [comment]
Hi Riad,

Nope, is this is prerequisite? Reviewing template and "bastion_bootstrap.sh" I understand that EIPs created by CF and should be assigned to instances by "bastion_bootstrap.sh" script during autoscaling. 

thanks

**Comment by yarrowfish** [comment]
Yes, you need to have it enabled otherwise it will fail during the AutoscalingGroup creation.  We had the same issue and we solved it by turning on "Enable auto-assign public IPv4 address".   I really think this needs to be documented as a prerequisite.  Try it and see if it works for you.  :-)

**Comment by strofimo** [comment]
It doesn't help in general. Still timeout on "AWS::AutoScaling::AutoScalingGroup" step. IP was assigned automatically. EIP is still not assigned. After fail I able to SSH to instance using auto-assigned IP. 
Still no cfn logs or "bastion_bootstrap.sh" script. But output of cloud-init is different now:

> Cloud-init v. 0.7.6 running 'modules:final' at Thu, 24 Aug 2017 11:46:38 +0000. Up 16.98 seconds.
> Downloading https://s3.amazonaws.com/cloudformation-examples/aws-cfn-bootstrap-latest.tar.gz
> error: [Errno 104] Connection reset by peer
> /var/lib/cloud/instance/scripts/part-001: line 14: cfn-init: command not found
> /var/lib/cloud/instance/scripts/part-001: line 15: cfn-signal: command not found

I don't know why connection reset. I tried to download this package with wget and it downloaded successfully. ,

**Comment by kamarja** [comment]
Hi,
Which OS are you using - Amazon Linux, CentOS, Ubuntu? I will try to reproduce the error.  EIP will not be assigned if there was an error in a previous step.

**Comment by strofimo** [comment]
I tried default option (Amazon Linux) and modified template when I explicitly set "ami-4fffc834" AMI.
I'm on US East(N. Virginia) region.

**Comment by kamarja** [comment]
Correct. It makes the call relative to the caller. 

I was able to launch without an issue in US-East (N. Virginia). Please try launching the template in a new VPC and use a different region. Just want to rule out the possibility of the issue being your existing VPC.

**Comment by strofimo** [comment]
I just tried Ubuntu 16.04 AMI and all works just fine in existing VPC. Let me try Amazon AMI again.

Side question - when I tried to connect to Ubuntu instance I see the banner but when:

> #                           Authorized access only!                           #
> #         Disconnect IMMEDIATELY if you are not an authorized user!!!         #
> #                All actions will be monitored and recorded.                  #
> ###############################################################################
> debug1: Authentications that can continue: publickey
> debug1: Next authentication method: publickey
> debug1: Trying private key: /root/.ssh/linux-bastion.pem
> debug1: key_parse_private2: missing begin marker
> debug1: read PEM private key done: type RSA
> debug1: Authentications that can continue: publickey
> debug1: No more authentication methods to try.
> Permission denied (publickey).

I don't have issues with connecting to instances with this private key before. Is bootstrapping bastion add restrictions I need to be aware of?

**Comment by strofimo** [comment]
Redone for Amazon and all works fine now. Not sure what it was, I saw these issues almost all week, 
I also was able to connect to instance through SSH. 
Thanks a lot for help!

**Comment by kamarja** [comment]
Thats awesome! No it does not add any restrictions. Make sure you are not using the wrong username ie: using ubuntu when it should be ec2-user etc.

**Comment by santiagocardenas** [comment]
@kamarja Great opportunity to add a few of the items above to the deployment guide as pre-requisites or notes:

- Requirement for auto-assign public IPs (required to communicate with AWS APIs during bootstrap)
- Username to use when connecting to bastion depending on AMI

**Comment by strofimo** [comment]
@santiagocardenas Regarding first pre-perquisite, why don't just add AssociatePublicIpAddress property to AWS::AutoScaling::LaunchConfiguration in template?

**Comment by Kostanos** [comment]
Hi, I got the same issue. It happens on running `cfn-init`:

```
[root@----- ec2-user]# cfn-init -v --stack "kostya3-bastion" --resource BastionLaunchConfiguration --region sa-east-1
Error occurred during build: Command b-bootstrap failed
```

Any idea what is missing here?

**Comment by Kostanos** [comment]
found issue in my case, the command b-bootstrap was misspelled:
```
            b-bootstrap:
              command:
                Fn::Sub: "./tmp/bastion_bootstrap.sh --banner ${BastionBanner} --enable ${EnableBanner} --tcp-forwarding ${EnableTCPForwarding} --x11-forwarding ${EnableX11Forwarding}"
```
instead of:
```
            b-bootstrap:
              command:
                Fn::Sub: "/tmp/bastion_bootstrap.sh --banner ${BastionBanner} --enable ${EnableBanner} --tcp-forwarding ${EnableTCPForwarding} --x11-forwarding ${EnableX11Forwarding}"
```

**Comment by santiagocardenas** [comment]
@strofimo Yes, we could do that!

@Kostanos That's interesting. Are the templates failing for you as they are right now in the master branch?

**Comment by VeryCrazyDog** [comment]
Knowledge sharing here. I encountered the same issue that the template timeout at `AWS::AutoScaling::AutoScalingGroup` step.

In my case, I was able to login to the Linux host and found the log at `/var/log/cloud-init*.log` and found that the script failed to download files from Internet. It is because we removed all traffic all source inbound rule in the Network ACL without adding back the [ephemeral ports](https://docs.aws.amazon.com/AmazonVPC/latest/UserGuide/VPC_ACLs.html#VPC_ACLs_Ephemeral_Ports).
